### ISSUE TITLE
Atomic<T> vs Untyped Atomics

### ISSUE THREAD

**Issue by cwfitzgerald** [gpuweb/gpuweb#2377]
## Context

This is a continuation of a discussion that was had on the WebGPU matrix. There has been some discussion of this problem in #2229 but was not the primary topic.

As of the writing of this issue, the WGSL spec defines all atomics as operating on dedicated Atomic<T> types. This is consistent with how MSL does atomics, but differs from how GLSL/SPIRV/HLSL/DXIL do atomics. They do "untyped" atomics where you can do an atomic operation on any T.

The purpose of this issue is to evaluate the issues with the current way WGSL does atomics and determine if WGSL could have "untyped" atomics.

## The Problem with `Atomic<T>`: Machine Translation from Other Languages

A very common path for the use of WebGPU and by extension WGSL, is through porting a project onto the web platform. These projects are written in some other source language and are being transpiled to WGSL. As it stands, such a machine translation might not be possible without conservatively promoting _too much_ to atomics.

Take the following GLSL:

```
layout(set = 0, binding = 0) buffer PrimeIndices {
    uint[] indices;
};
```

#### How should that be translated to WGSL if a static index 15 is accessed?

It could promote the entire array to atomics `array<atomic<u32>>` which would pessimise other reads.

It could pull some heroics and split the array into 3 parts, a `array<u32, 15>` header, a `atomic<u32>` middle and a `array<u32>` footer.  This would over-promote anything that accessed 15 non-atomically, but leave the rest of the array unscathed. This messes with the minimum binding size and is definitely something I consider "crazy" to expect implementations to do.

#### How should that be translated to WGSL if a dynamic index is accessed?

The translator would have no choice but to over-promote to `array<atomic<u32>>`. 

## Can We Do Untyped Atomics

On all backends with the exception of MSL, we can do untyped atomics just fine. On metal, in order to do an atomic operation you need to have that explicit type.

### Casting

When presented with an atomic, spirv-cross translates this to an explicit cast from the non-atomic type to the atomic type.

See https://shader-playground.timjones.io/6ac47edae3a6940746a1b92e06dc7836 for the full source, but the important bit is that it translates:

```glsl
atomicExchange(indices[0], 2)
```

to

```msl
atomic_exchange_explicit((device atomic_uint*)&_10.indices[0], 2u, memory_order_relaxed);
```

This kind of re-interpretation seems a bit sketchy at first but is not without president within Apple's own example code. In their example code [Modern Rendering with Metal](https://developer.apple.com/documentation/metal/modern_rendering_with_metal) they don't _cast_, but they reinterpret that same threadgroup variable as both a `atomic_uint*` (Renderer/Shaders/AAPLLightCulling.metal:551) and a `float*` (Renderer/Shaders/AAPLLightCulling.metal:531).

### Mixing Atomics and Non-Atomics

The question becomes, if we determine that untyped atomics are a thing we want, how should we describe the behavior if the user mixes and matches atomics and non-atomics?

I don't have a solid formal answer to this, but logically this feels no different from racing non-atomic reads and non-atomic writes across a RW buffer without barriers. In the case of racing non-atomics, you end up with some unspecified value, possibly an old value, possibly reading a partial write. We don't want to ban that for obvious reasons, so this would be no worse than the situation there. 

Trying to get around this by statically defining what is an atomic (through Atomic<T> or validation) is just as overly restrictive as trying to solve the problem of the racing non-atomics and has similarly unspecified results.

## Conclusion

I have tried to express a route that we could make untyped atomics work in WGSL. I think this is particularly important as it falls into the category of "preserving developer trust". For years before they could run on metal and use untyped atomics through spirv-cross, but us coming in and deciding that you can't do that, is the kind of thing that would really put people off of WebGPU.

If there are any questions or if anything needs clarification, please ask. I can also edit the post to include other pointers pro/con.

**Comment by alan-baker** [comment]
@litherum for comments on casting in Metal.

**Comment by alan-baker** [comment]
For reference, this was discussed in https://github.com/gpuweb/gpuweb/issues/1360#issuecomment-779985393 as part of the original proposal.

**Comment by litherum** [comment]
Yeah, that comment you linked to is correct. Casting in the general case would be no bueno, and mixing atomics and non-atomics would be the same. This isn't a situation like depth textures; we really need this type for atomics to work at all.

**Comment by dneto0** [comment]
I sympathize with people seeing this as a loss of functionality, coming from other platforms to WebGPU.  It is a loss.

Using casting tricks may trick the tools into letting you do what you want, today, on shipped implementations.   But that's not the programming model promised by the stack.  The MSL constraints are inherited from  C++14, and what you're asking for is non-starter in C++14.

There was a somewhat similar situation where I very much wanted the ability to form a pointer to a component in a vector.  After close reading of C++14, I found this was not allowed because it violates the type-based alias analysis rules.
See the discussion here:  https://github.com/gpuweb/gpuweb/issues/1931#issuecomment-953180418 

In that case, an MSL vector type does not count as an "aggregate" or "union", which meant you couldn't write the memory as one type (e.g. vector) and read it as another (e.g. vector component).  Similarly, a C++ atomic type does not count as an aggregate or union over its contained integer type.   So the whole implementation stack (compiler and machine) is not built to do what you're asking.

About "preserving developer trust".  That only counts or has meaning when you've already restricted yourself to features that are available in WebGPU itself.  In this case, what you're asking is outside what WebGPU can do.

Now, it would be great if you could convince Apple to add the feature to a supported shader language.  Then I'd be first in line to enable it in WGSL and WebGPU.

**Comment by m-schuetz** [comment]
This is unfortunate because at least for point cloud rendering in OpenGL, mixing non-atomic reads with atomic writes can be an enormous performance boost (up to 4x). The non-atomic reads allow us to do an early-depth test, and thereby avoid a large amounf of superflous atomic writes. Unfortunately I wasn't able to check if atomic loads are a suitable replacement in our case since I couldn't get atomicLoad to work in OpenGL.

**Comment by m-schuetz** [comment]
This is unfortunate because at least for point cloud rendering in OpenGL, mixing non-atomic reads with atomic writes can be an enormous performance boost (up to 4x). The non-atomic reads (from potentially cached allow us to do an early-depth test, and thereby avoid a large amount of superflous atomic writes. Unfortunately I wasn't able to check if atomic loads are a suitable replacement in our case since I couldn't get atomicLoad to work in OpenGL. 

I'm mostly afraid that atomic loads would force global memory access. However, the atomicLoad extension seems to support specifying the scope within which the call is atomic. Is this something that could be done in WebGPU?

[1] https://github.com/KhronosGroup/GLSL/blob/master/extensions/khr/GL_KHR_memory_scope_semantics.txt

**Comment by dneto0** [comment]
The GL_KHR_memory_scope_semantics extension exposes the Vulkan memory model to GLSL.  It's not supported on OpenGL, which is explains why atomicLoad doesn't work in OpenGL.

> I'm mostly afraid that atomic loads would force global memory access.

Yes, atomic loads will force an access that goes all the way out to the authoritative copy of the memory, i.e. past all the caches.

> However, the GLSL atomicLoad extension seems to support specifying the scope within which the call is atomic. Is this something that could be done in WebGPU?

See https://gpuweb.github.io/gpuweb/wgsl/#scoped-operations, which explains how the WGSL atomics map to Vulkan memory model concepts.
In particular:
 - an atomic on a 'storage' buffer (SSBO) uses a memory scope of QueueFamily. 
 - an atomic on workgroup uses a memory scope of Workgroup.

Also, WGSL atomics are always relaxed.

If buf.x is in a storage buffer, then WGSL's

     atomicLoad(&buf.x)

is the same as the GLSL:

     atomicLoad(buf.x, gl_ScopeQueueFamily, gl_StorageSemanticsBuffer, gl_SemanticsRelaxed  );

If buf.x is in workgroup, then the GLSL would be:

     atomicLoad(buf.x, gl_ScopeWorkgroup , gl_StorageSemanticsShared, gl_SemanticsRelaxed  );

**Comment by alan-baker** [comment]
MSL and HLSL don't support specifying atomics with fine-grained scoping like you're asking for. In order to implement the same behaviour everywhere, the scope needs to be based on storage class. So while we could spec an atomicLoad that had a workgroup memory scope for storage buffers, it would still be implemented by a device atomic in HLSL and MSL.

**Comment by litherum** [comment]
I think I was a bit too curt in https://github.com/gpuweb/gpuweb/issues/2377#issuecomment-985795220. The reason is that it’s undefined behavior to read/write using atomic operations and then read/write using non-atomics. The type system is a way to enforce that this never happens, and this must be enforced for the security guarantees of WGSL.

**Comment by Kangz** [comment]
> The reason is that it’s undefined behavior to read/write using atomic operations and then read/write using non-atomics.

Note that this would be undefined behavior in other shading languages too. However undefined behavior there isn't "anything goes" but that the exact values you read / write to memory aren't completely defined (but they can't be leftover data from someone else).

> this must be enforced for the security guarantees of WGSL.

Can you explain why this is important to uphold the security guarantees of WGSL? The WebGPU synchronization scopes should mean that the only undefined values you load / write are the page's own value, not someone else's values. So while undefined, it is safe undefined behavior no?

**Comment by m-schuetz** [comment]
If Metal seems capable of casting/reinterpreting the array from atomic to non-atomic in one of their examples, would it be feasible to do the reverse in WebGPU? Reinterpreting an atomic memory location as a non-atomic memory?

**Comment by kvark** [comment]
@dneto0 
> Yes, atomic loads will force an access that goes all the way out to the authoritative copy of the memory, i.e. past all the caches.

Could you explain this bit? We are talking about fragment shaders here, there are no synchronization barriers within a render pass. So what forces the driver to go all the way to the "authoritative copy"? No sync means any value from the cache should be fine (i.e. correct according to guarantees we can provide), as long as the cache was populated in this render pass.

**Comment by m-schuetz** [comment]
Does that also hold for compute shaders? Because that's where I'd like to mix atomic/non-atomic access.

**Comment by kvark** [comment]
Unless you have a `storageBarrier`/`workgroupBarrier` between a store and a load, I don't see why the driver would need to force its way into the global/workgroup memory, respectively.
This is a bit speculative. I'm asking @dneto0 to help with building this kind of intuition.

**Comment by m-schuetz** [comment]
> Unless you have a storageBarrier/workgroupBarrier between a store and a load, I don't see why the driver would need to force its way into the global/workgroup memory, respectively.

>  "workgroup atomics have a Workgroup memory scope in SPIR-V, while storage atomics have a QueueFamily memory scope in SPIR-V."

QueueFamily seems to be a pretty far-reaching scope already? Compared to workgroup scope.

**Comment by dneto0** [comment]
>> Yes, atomic loads will force an access that goes all the way out to the authoritative copy of the memory, i.e. past all the caches.

> Could you explain this bit? We are talking about fragment shaders here, there are no synchronization barriers within a render pass. So what forces the driver to go all the way to the "authoritative copy"? No sync means any value from the cache should be fine (i.e. correct according to guarantees we can provide), as long as the cache was populated in this render pass.

My language is a bit loose. "authoritative" is meant to be suggestive.  So much depends on how the architecture of the GPU, but the key thing is that the fundamental contract of atomics have to be satisfied.

The key properties of atomic variables **in C++** are:
- All modifications to a particular variable are totally ordered.  Any observer will witness a subset of that total order.
     - Note: There is no requirement that the total orders for different atomic variables be consistent in a global or causal sense.
- A modification can't be "torn": each time an observer witnesses the state of the atomic variable, it's always the state of the variable contents **between** such modifications. (E.g. you don't see the top 16bits from modification M and the bottom 16bits from modification M+1.)

For GPU code you have to be more careful about "total order" because that only applies for accesses during execution of the shader.  There are host writes before the shader starts (to initialize the variable), and maybe host reads after the shader ends.  Also, Vulkan allows mixing non-atomic accesses with atomic accesses (provided the appropriate synchronization is in place).  So it refines "total order" to "mutually-ordered atomics", where that occurs under circumstances described in the Vulkan memory model.

And the corresponding rule about atomics ordering observability in Vulkan is the [scoped-modification-order](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#memory-model-scoped-modification-order):

> For a given atomic write A, all atomic writes that are mutually-ordered with A occur in an order known as A’s _scoped modification order_. A’s scoped modification order relates no other operations.

Ok, so "authoritative copy of memory" in what I wrote above means "use whatever mechanism is required to provide as-if scoped-modification-order observations to non-racy observers".  That only has to be satisfied between mutually ordered atomics, which itself is at most the lifetime of the shader.

If the atomic variable is in an SSBO, then the lifetime of the variable is the whole render pipeline for the whole draw call. Every fragment in the render pipeline can perform an atomic operation on that variable.  Assuming the fragment shaders don't have races, then all the atomic modifications can be mutually ordered, and so however you organize the memory physically must respect the scoped-modification-order constraint.  The GPU architecture might be clever and put that in a cache, provided it can offer the guarantee (e.g. lots of serialization and pipelineing, but don't do very much in parallel).

For compute shaders you have the same constraint when the atomic variable is in an SSBO. If you have lots of workgroups in parallel, interacting with an atomic variable in SSBO, then you can have expensive communication.  If the implementation limits itself to running one workgroup at at time, then the SSBO contents can be cached very close to the workgroup execution units and it can run very fast.
However, if the atomic variable is in workgroup memory, then each workgroup always gets its own copy of that variable and communication is limited to within a single workgroup. That ought to be much faster.

**Comment by dneto0** [comment]
> QueueFamily seems to be a pretty broad scope already? Compared to workgroup scope.

Yes, QueueFamily is very big, and can span what you would normally think of a device, at least from the shader's perspective. So it's much bigger than a workgroup.

Stack overflow has a good answer on what a queue family really is. https://stackoverflow.com/questions/55272626/what-is-actually-a-queue-family-in-vulkan

**Comment by dneto0** [comment]
> Unless you have a storageBarrier/workgroupBarrier between a store and a load, I don't see why the driver would need to force its way into the global/workgroup memory, respectively.
This is a bit speculative. I'm asking @dneto0 to help with building this kind of intuition.

With proper use of atomics, they always see authoritative copies, as per above.
Atomics translate to slightly tweaked memory access instructions in the shader; they are specific to each GPU.  The driver doesn't actively intervene, but the total system design determines how expensive those operations are.

Non-atomic accesses need extra help so you can see sensible data.
The barriers are used to constrain accesses to be ordered before and after the barrier. (Typically you mostly care about the non-atomic accesses)
The pattern in the shader is:

 1. Do a bunch of writes to memory (workgroup or storage buffer). These are typically non-atomic.
 2. Execute a barrier
 3. Read the memory that was written in step 1.  These are typically non-atomic.  Invocations here will "see" the results when the writes came from different invocations.

Normally, non-atomic reads and writes can be very lazy. Writes don't poke out all the way to "authoritative" memory: they can be cached locally, waiting to be pushed out.  Reads don't reach out all the way to the "authoritative" memory: they only go as far as a local cache.

The barriers are the explicit request by the programmer to "push" the prior writes out to where they will become authoritative, and "pull" the authoritative copy down more closely (or more often invalidate the local caches).
The programmer tells the system when to do that. It's not a driver-level thing, but it's extra instructions in the shader (specific to the GPU).  (Vulkan models the "push" generically as an "availability" operation, and the "pull" generically as a "visibility" operation.)

The many-writes, barrier, many-reads pattern is useful because algorithms can be designed to operate in phases that way, and it amortizes the cost of the cache management.  So with a little extra burden on the programmer to express themselves, it's a big win.

**Comment by raphlinus** [comment]
For what it's worth, I'm in favor of this proposal. I believe the best path forward is for Metal to redefine mixed atomics and non atomics as non-undefined behavior. That would have the additional benefit of supporting the widespread existing usage of spirv-cross.

I believe inheriting undefined behavior from C++14 in bulk is a mistake, and a departure from the original stated goals of being consonant with SPIR-V semantics. One other area I've found is signed integer overflow. If we are serious about respecting C++14 UB, then that forces translation to generate code sequences for signed integer arithmetic that check for overflow, which is likely to have serious performance consequences. I haven't audited the various sources of UB in C++14 but wouldn't be surprised if there were others.

Thus, I believe that in order to ship a usable WGSL, it is necessary to be flexible about existing UB in MSL. The question is where to draw the line. I think the advantages of drawing it to include mixed atomics outweigh the disadvantages.

**Comment by dneto0** [comment]
> I believe the best path forward is for Metal to redefine mixed atomics and non atomics as non-undefined behavior.

That would be great.  But it's not something W3C can resolve.   I don't know how Apple defines and executes its Metal roadmap, but you'd have to send your feedback via those channels somehow.

(I personally think it's quite good to have "this important thing X is possible on D3D and Vulkan but not Metal!" blog posts from serious users such as you. But I don't make decisions for Apple.)

> One other area I've found is signed integer overflow. 

We've designed WGSL integer arithmetic semantics so you can always fall back on unsigned integer arithmetic (uint32_t).
(There is still an outstanding issue with division, but the committee just needs to pay attention for a bit.)
Tint handles those translation for you. For details, see https://bugs.chromium.org/p/tint/issues/detail?id=124

**Comment by dneto0** [comment]
I don't think this can be accommodated in the first version of WebGPU.
Suggest pushing this for consideration post-v1.

Ordinarily I would suggest an extension to enable the feature, but then that would leave out Apple platforms and fragment the web platform.

**Comment by raphlinus** [comment]
Ah, I didn't realize the bitcast-to-unsigned translation strategy was viable. I'm pretty sure that spirv-cross doesn't do any such thing, but at the same time I understand that the requirements are a bit different. Good to know the safety considerations are being taken seriously!

**Comment by cwfitzgerald** [comment]
Right now in WebGPU, you can have any of the following conditions happen when you do an non-atomic read from a read-write buffer that another thread was writing to non-atomically:
- You read the current value
- You read the value from some time in the past.
- You read a partial write, leaving you with garbage.
- You read complete garbage.

Note this never includes values that are derived from a previous sync-scope (or user's) data. This means that, despite the value you are reading being completely undefined, it is perfectly safe and as such WebGPU allows it.

I think the core of the issue is _why_ casting non-atomic -> atomic in MSL is a very bad thing. If it is bad because of any of the above reasons (you read potentially arbitrary junk depending on cache state and a million other factors), this is perfectly within what WebGPU has considered okay.

@litherum thoughts?

**Comment by raphlinus** [comment]
Thanks @cwfitzgerald for bringing up data races. Those are a much better example of undefined behavior that we do not want to inherit from C than signed arithmetic. To me, mixing atomics and non-atomics is even more benign than a data race, but it's possible I'm missing something.

That said, I'd like to offer a proposal that might satisfy the concern of automatic translation of shaders that use atomics, but in a different way than weakening the type system. Instead of trying to facilitate both atomic and nonatomic access to the same memory location, keep it typed as atomic but allow weak semantics as an option to atomic load/store operations. One precedent for this idea is the `.weak` qualifier on load/store in the [PTX memory model](https://github.com/NVlabs/ptxmemorymodel/blob/master/PTXMemoryModelASPLOS2019.pdf). I believe this also corresponds to the [unordered](https://llvm.org/docs/Atomics.html#unordered) atomicity level in LLVM, and is used to model potential races in the Java memory model; likely a good source of inspiration for the goals of WebGPU, where C-style UB is undesirable.

I would define this so "weak" (or "unordered") semantics map to OpLoad and OpStore in SPIR-V. My read of the [data race](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#memory-model-access-data-race) section of the Vulkan memory model is that these non-atomic operations may be mixed with atomics to the same location as long as there is additional synchronization, in other words a similar expectation as mixing non-atomic reads and writes. I'd be happy to be corrected if I'm wrong.

An additional potential refinement is that we could have an even weaker "private" atomic semantics that would not be subject to barriers (or acquire/release semantics in a future extension). Assume for the sake of argument we go with Approach 1 of #2228, in which the default for load/store operations is to set the NonPrivatePointer flag (and possibly other flags for availability/visibility). My proposed ultra-weak semantics would map to OpLoad/OpStore _without_ the NonPrivatePointer flag set.

The above mapping to SPIR-V is what I propose as the normative mapping. I think it's also worth mentioning other mappings. The intent of these weak semantics is to generate ISA that reads from local cache on cache hit (for a load; similar idea for stores). In particular, on AMD it would generate buffer_load_dword without the DLC/GLC bits set. All this is documented in the [AMDGPU backend for LLVM](https://llvm.org/docs/AMDGPUUsage.html#amdgpu-amdhsa-memory-model-code-sequences-gfx10-table) and is a natural consequence of mapping "relaxed" to LLVM "monotonic" and the new weaker semantics to "unordered."

My understanding of HLSL is that it is similar to SPIR-V; since there are no explicit atomic load and store operations, mixing of Interlocked and non-atomics is common practice, and if there were a spec, it would likely respect that practice.

Metal is a slightly different question. In my mind, there are two relevant dialects, which we might call "Metal as she is spoke" and "the Queen's Metal." In the former (which corresponds to spirv-cross output among other things), a reasonable translation is `*(uint32_t *)atomic_ref`, but that undoubtedly sounds crude to the ears of posh Metal speakers. For the latter, upgrading to relaxed semantics is certainly viable, possibly at some performance cost. I would personally consider this to be an internal question of a potential Metal specification effort and not necessarily the business of WebGPU; the important thing is that there *is* a viable translation.

Lastly, I think this discussion would benefit greatly from some empirical data on the actual performance implications of such fine-grained control over weak semantics. My gut feeling is that there is probably a use case and GPU for which such control would make a significant, and possibly dramatic difference, but I don't have such a case in hand. I'm doing a fair amount of work at the moment in porting and performance characterization, so I will keep my eyes and ears open. I think this is also a good opportunity for somebody to do something analogous to Seb Aaltonen's [perftest](https://github.com/sebbbi/perftest/blob/master/README.md) but with a focus on atomics and coherence.

**Comment by AmesingFlank** [comment]
> A very common path for the use of WebGPU and by extension WGSL, is through porting a project onto the web platform. These projects are written in some other source language and are being transpiled to WGSL. As it stands, such a machine translation might not be possible without conservatively promoting too much to atomics.

This is a very good point. The [very first goal](https://www.w3.org/TR/WGSL/#goals) listed in the WGSL spec is for WGSL to be "Trivially convertable to SPIRV". Having to conservatively over-promote entire buffers is by no means a trivial transformation.

**Comment by kvark** [comment]
Filed FB9807778 to clarify the situation about the titles that have already shipped, and with a question of whether or not aliasing the same memory location consistently (across an entry point) is still UB.

**Comment by dneto0** [comment]
> Right now in WebGPU, you can have any of the following conditions happen when you do an non-atomic read from a read-write buffer that another thread was writing to non-atomically:

... followed by an explicit list.

Similarly

>  However undefined behavior there isn't "anything goes" but that the exact values you read / write to memory aren't completely defined (but they can't be leftover data from someone else).

No. Everybody stop.  Undefined behavior means **the program has broken the contract with the system** and therefore **all bets are off**, meaning the system can behave entirely contrary to the programmer's expectation.  (External considerations for the browser means you should have a level of sandboxing to contain the damage.)

For background, and how it ties to WebGPU, see [my presentation](https://docs.google.com/presentation/d/1D2ZmqMa3GQJ13xcUMfaWmIJDv4AwZwUGB6fIoLLaCXQ/edit#slide=id.p4) from a prior WebGPU F2F.

Metal has a contract with the programmer.  It is documented as mixture of C++14, and extra prose in various manuals. That contract says: you shall not mix atomic and non-atomic accesses on the same locations. If you do, you have undefined behaviour:  unicorns and rainbows, printer on fire, that kind of thing.

No amount of arguing "but it works on X today" gets around the fact that Apple has explicitly taken the opposite stand.
This gives them freedom of implementation, to optimize future implementations (combinations of hardware and software).  They clearly are exercising that option, by making their own portfolio of CPU/GPU devices, and will continue to do so.
(Good for them!)

Apple's implementation is based on the C++ compilation flow through Clang and LLVM.  Asking them to promise more than C++ does in this regard is a large request, you may as well take it to the ISO C++ committee and try your luck there. The LLVM stack aggressively exploits undefined behaviour, and you're asking Apple to promise, now and forever, more than what their compiler stack is promising, i.e. it's effectively a fork.

I'd be ecstatic if we could make progress by convincing Apple to change course on this issue.  But I will not hold up WebGPU v1 for this. (Hence I pushed this issue to post-v1.)

Aside: 
HLSL also has a contract with the programmer. It is somewhat documented, but somewhat more folkloric and based on a large amount of existing practice. There's a very strong feedback loop between the implementation and the extremely broad catalog of content running on DX*, and that establishes the agreed-upon semantics here.  (Not to denigrate Microsoft here: they've built an amazingly successful ecosystem. Kudos.)

Aside: GLSL and OpenGL were/are also in the same murky bucket as HLSL/D3D.  Vulkan and SPIR-V's work in this area was a first-principles revisitation, hand in hand with formal modeling and empirical testing. The formal modeling sharply clarifies the questions you can pose and the answers you can get with confidence.

**Comment by dneto0** [comment]
Hm.  Shorter way of saying it:  "works on NVIDIA" is the new "works on x86", and similarly insufficient.

**Comment by raphlinus** [comment]
> Metal has a contract with the programmer. It is documented as mixture of C++14, and extra prose in various manuals.

Could you or someone from Apple point me to the prose that explicitly allows data races (in the Java memory model sense, treating them as "unordered" in LLVM language) by non-atomic memory operations? That would be enormously clarifying for this discussion. A search of the [Metal Shading Language](https://developer.apple.com/metal/Metal-Shading-Language-Specification.pdf) doc for the word "race" yields only "Objects of atomic types are the only MSL objects that are free from data races."

**Comment by dneto0** [comment]
> A search of the [Metal Shading Language](https://developer.apple.com/metal/Metal-Shading-Language-Specification.pdf) doc for the word "race" yields only "Objects of atomic types are the only MSL objects that are free from data races."

Right.  And so the remaining rules about races and the memory model are inherited from C++14.

I'm no expert in the Java memory model.  I would resort to a search.  E.g. https://rsim.cs.uiuc.edu/Pubs/popl05.pdf  But I have no idea how that paper is situated in the broader literature (both Java and non-Java), or its relation to machines since 2005.

**Comment by kvark** [comment]
Side note. If we are talking about unsynchronized access to the same memory location as both atomic and non-atomic, then we already allow this in WebGPU, just not within the same shader. I.e. you can have a render pass with 2 draw calls. First draw call uses a fragment shader with
```rust
struct Foo {
  a: atomic<u32>,
};
[[group(0), binding(1)]]
var<storage, read_write> buffer: Foo;
```
The second draw call uses a fragment shader with:
```rust
struct Foo {
  a: u32,
};
[[group(0), binding(1)]]
var<storage, read_write> buffer: Foo;
```

Since the whole render pass is a single synchronization scope, these accesses may easily happen at the same time from both fragment programs.

---

This actually matches something @dneto0 expressed on the office hours call: the semantics of memory access here is not the main problem. The very fact we are casting in a shader is the problem.

**Comment by litherum** [comment]
Before I start talking about atomics, it's helpful to make a simpler example that doesn't use atomics.

## Simple example

Consider a WGSL program like this:

```
struct ABuffer {
    index: u32;
    result: u32;
    anArray: array<u32, 2>;
}
[[group(0), binding(0)]] var<storage> aBuffer : ABuffer;

[[stage(compute)]] fn foo() {
    aBuffer.anArray[aBuffer.index] = 2;
    aBuffer.anArray[aBuffer.index] = 3; // Middle line
    aBuffer.result = aBuffer.anArray[aBuffer.index];
}
```

Let's transcribe this into MSL:

```
struct ABuffer {
    uint32_t index;
    uint32_t result;
    uint32_t anArray[2]
};

kernel void foo(device ABuffer* aBuffer [[buffer(0)]]) {
    aBuffer.anArray[aBuffer.index] = 2;
    aBuffer.anArray[aBuffer.index] = 3; // Middle line
    aBuffer.result = aBuffer.anArray[aBuffer.index];
}
```

After running this program, you'd expect aBuffer.result to hold 3. Easy peasy.

Now, let's tweak that MSL a bit. Let's change the middle line to do a `float` store instead of a `uint32_t` store.

```
struct ABuffer {
    uint32_t index;
    uint32_t result;
    uint32_t anArray[2]
};

kernel void foo(device ABuffer* aBuffer [[buffer(0)]]) {
    aBuffer.anArray[aBuffer.index] = 2;
    *reinterpret_cast<device float*>(aBuffer.anArray + aBuffer.index) = 3; // Middle line
    aBuffer.result = aBuffer.anArray[aBuffer.index];
}
```

After running this program, `aBuffer.result` holds 2! (I observed this on my main desktop machine.) How can that be?

One of the steps that a compiler does when it analyzes a function like this is it tries to eliminate loads and stores. In the first line of the program, it sees a store to memory, and in the last line of the program, it sees a load of that same address. The compiler is going to try to eliminate the load (because of course it will).

In order to do this, the compiler has to look at all the code between the store and the load, to see if anything conflicts. Here, "conflicts" is a word I just made up, but it does have a rigorous meaning - the C++ spec has a list of what it means for a load and a store to conflict. If something in the middle conflicts, that means bummer, we can't eliminate the load; but if nothing conflicts, then we can safely eliminate the load no problem. One of the requirements to determine whether or not a memory operation can conflict is that the operations have to have the same type (for some (rigorous) definition of "same").

In the first MSL example, the middle line's store has the same type as the last line's load, so that means they conflict; we can't eliminate the load. (We can, however, eliminate the load if we run the same analysis but consider the middle line and the last line together, rather than the first line and the last line together.) In the second MSL example, the middle line's store doesn't have the same type as the last line's load, so they don't conflict, and we can "safely" eliminate the load. Ruh-roh.

## Atomics

Now, of course, no one is proposing that we turn regular loads and stores into that `reinterpret_cast` gunk above. However, I believe the OP _is_ proposing doing that with atomics. So, if the author replaces the middle line with an atomic store (or an atomic CAS - it works just the same), the generated Metal would look like this (let's call this "listing 1"):

```
struct ABuffer {
    uint32_t index;
    uint32_t result;
    uint32_t anArray[2]
};

kernel void foo(device ABuffer* aBuffer [[buffer(0)]]) {
    aBuffer.anArray[aBuffer.index] = 2;
    atomic_store_explicit(reinterpret_cast<device atomic<uint32_t>*>(aBuffer.anArray + aBuffer.index), 3, memory_order_relaxed); // Middle line
    aBuffer.result = aBuffer.anArray[aBuffer.index];
}
```

This has the same problem as the above. The `atomic<uint32_t>` type doesn't conflict with `uint32_t`, so the compiler is free to write 2 into `aBuffer.result`.

## Bounds checks

Consider the situation where `aBuffer.anArray[aBuffer.index]` was used to index into yet another array. That access to the other array would be, of course, guarded by a bounds check:

```
    // These 2 lines are copied directly from above
    aBuffer.anArray[aBuffer.index] = 2;
    atomic_store_explicit(reinterpret_cast<device atomic<uint32_t>*>(aBuffer.anArray + aBuffer.index), 3, memory_order_relaxed);

    // Use aBuffer.anArray[aBuffer.index] to index into a bounds-checked array
    uint32_t temp = aBuffer.anArray[aBuffer.index];
    if (temp > arrayLength) { // Bounds check
        ... do something sensible ...
    } else {
        ... = someBuffer[temp]; // Access
    }
```

1. Putting `aBuffer.anArray[aBuffer.index]` into a variable doesn't save you - there's no guarantee that anything you put into a variable will actually be held in a register for its lifetime.
2. There's no guarantee that the `temp` in the bounds check won't have its load eliminated using the same reasoning as before. Also, the `temp` in the access might _not_ have its load eliminated, because the compiler is not required to eliminate loads.
3. Using `min()` instead of `if` doesn't save you, because there's no guarantee that `min()` isn't implemented with a conditional, and isn't inlined.

So this is an example where some innocuous-looking WGSL code could end up defeating a MSL bounds check.

## Current Investigations

It turns out, if you actually run the "listing 1" example above (on 2 out of 2 devices I tried), the right thing happens - `aBuffer.result` has a value of 3 after the program is finished executing. But, this isn't because the compiler understands that `atomic<uint32_t>` conflicts with `uint32_t`. This is because of the specific operation being used: `atomic_store_explicit()`. On at least some devices, the compiler pessimizes, and says "I'm going to treat that `atomic_store_explicit()` call as if it might have trampled over all memory, so we'd better act as if everything after the call needs to be reloaded from memory." No specification says that this pessimization is required - though it is, of course, allowed.

Therefore, from a purely spec point of view, supporting HLSL-style atomics in MSL is totally allowed to defeat bounds checks, though on at least one sample program running on two devices, the bounds checks wouldn't be defeated in practice because of compiler pessimizations.

I cannot claim that the pessimization will exist forever, on any devices, on any shader generated by any WGSL -> MSL compiler. (Of course I cannot claim that - I cannot make claims about the future at all.) I am, however, trying to get more information about this pessimization - it's conceivable that, after further investigations, we might be comfortable with baking in this pessimization into the design of WGSL. We're still investigating.

**Comment by AmesingFlank** [comment]
Thanks @litherum for the recount of memory order issues. I'd like to mention that MSL supports memory fences, see for example section 6.9.1 in the [MSL spec](https://developer.apple.com/metal/Metal-Shading-Language-Specification.pdf):
> Table 6.12 describes the bit field values for the mem_flags argument to
threadgroup_barrier and simdgroup_barrier.
> ...
> mem_device: Ensure correct ordering of memory operations to device memory.
It seems to me that these fences would be sufficient to ensure that the compiler will operate pessimistically and reload everything from the memory, and a MSL implementation for WGSL could use this to support atomics. Wdyt?

**Comment by litherum** [comment]
Atomics can be used from non-uniform control flow, but barriers can't, so I don't think those fit the need. Also barriers are only valid in computer shaders.

**Comment by kvark** [comment]
That's a very nice write-up, thanks @litherum ! It gets straight to the point of being a potential security hazard.
One part of it is bothering me though, as someone not specializing in compiler tech:
```cpp
uint32_t temp = aBuffer.anArray[aBuffer.index];
```
> Putting aBuffer.anArray[aBuffer.index] into a variable doesn't save you - there's no guarantee that anything you put into a variable will actually be held in a register for its lifetime.

Naively, I'd thought that the compiler can do whatever, but it needs to preserve the appearance of "temp" as a single value, since it's not mutated after this assignment. But what you are saying is that (and please correct me if that's a misunderstanding!):
  1. a single variable can in fact manifest itself with different values
  2. it's only possible because the compiler wasn't able to reason about the mutability of the memory location that's the source of this value
  3. this happened only because we ruined compiler guarantees with this reinterpret cast (type-based aliasing analysis, TBAA), entering the UB territory

Did I get this right?
In this case, would MSL's C++ have a way to mark this memory location as volatile? Something like
```cpp
void safe_atomic_store(volatile device uint32_t* pointer, uint32 value) {
  atomic_store_explicit(reinterpret_cast<device atomic<uint32_t>*>(pointer), value, memory_order_relaxed);
}
...
    aBuffer.anArray[aBuffer.index] = 2;
    safe_atomic_store(&aBuffer.anArray[aBuffer.index], 3)   
    uint32_t temp = aBuffer.anArray[aBuffer.index];
```
So the compiler sees that the memory location of this type `uint32_t` is seen as a `volatile` pointer first (of the same type!), which could convince it that it shouldn't rely on the old value. It's just a thought, I haven't tried it.

**Comment by litherum** [comment]
Oh yeah, I forgot to say: volatile doesn't save you, because it just forces reads and writes to that variable to happen in order, but doesn't provide ordering guarantees for _other code_ relative to the reads and writes on the volatile variable.

In the "reinterpret_cast<device float*>" example, the load in the last line loads a "2", which is as-if the load had happened before the store on the middle line of the program. If the middle line stored to a volatile device float*, then that would guarantee that the store _happened_, but wouldn't guarantee that the next line of the program executed as-if it occurred after the middle line.

**Comment by kvark** [comment]
@litherum hold on, I've been thinking about this some more and realized why this logic is incorrect.

Supposing what I described under "Did I get this right?" is correct. The compiler appears to need to know about all loads and stores (or other accesses) to a memory location used with a particular type, in order to optimize properly. This is wrong because we are programming a single thread of many, and are accessing a shared resource. This read-write storage buffer can be written to by entirely different shaders at the same time the current one is being executed. Therefore, if we assume the compiler logic as described, even a simple bound check (without any stores!) would possibly fail:
```cpp
    uint32_t temp = aBuffer.anArray[aBuffer.index];
    if (temp > arrayLength) { // Bounds check
        ... do something sensible ...
    } else {
        ... = someBuffer[temp]; // Access
    }
```

Clarification on why it would fail. @litherum point was that the compiler is free to replace `temp` value accesses with loads. If it does that, the loads may pick up different values based on what's written to the buffer by other threads executing concurrently.

TL;DR: I don't think atomic aliasing makes the security story any worse than it is right now. Either the compiler sees storage buffers as completely volatile (therefore not messing up our load/stores), or we can't make generated MSL code secure in principle, and we can all go home.

**Comment by jrmuizel** [comment]
Is it possible to just use a `union` instead of a `reinterpret_cast` to let TBAA know what's going on?

e.g.
```c
union uf {
        float f;
        uint32_t u;
};

kernel void foo(device ABuffer* aBuffer [[buffer(0)]]) {
    aBuffer->anArray[aBuffer->index] = 2;
    reinterpret_cast<device uf*>(aBuffer->anArray + aBuffer->index)->f = 3; // Middle line

    aBuffer->result = aBuffer->anArray[aBuffer->index];
}
```
compiles to
```llvm
define void @foo(%struct.ABuffer addrspace(1)* noalias nocapture %0) local_unnamed_addr #0 {
  %2 = getelementptr inbounds %struct.ABuffer, %struct.ABuffer addrspace(1)* %0, i64 0, i32 0
  %3 = load i32, i32 addrspace(1)* %2, align 4, !tbaa !20
  %4 = zext i32 %3 to i64
  %5 = getelementptr inbounds %struct.ABuffer, %struct.ABuffer addrspace(1)* %0, i64 0, i32 2, i64 %4
  %6 = bitcast i32 addrspace(1)* %5 to float addrspace(1)*
  store float 3.000000e+00, float addrspace(1)* %6, align 4, !tbaa !25
  %7 = getelementptr inbounds %struct.ABuffer, %struct.ABuffer addrspace(1)* %0, i64 0, i32 1
  store i32 1077936128, i32 addrspace(1)* %7, align 4, !tbaa !26
  ret void
}
```

which is the intended result.

**Comment by cwfitzgerald** [comment]
FWIW, union aliasing is technically UB in C++, but it's UB that basically every codebase has, so has a very good chance of "never being used" in optimization.

**Comment by raphlinus** [comment]
I'm very much looking forward to the result of the exploration on the Metal side. In the meantime, I have more information from the C++ side, which might be useful.

As of C++20, the language has exactly the mechanism we need, in the form of [atomic_ref](https://en.cppreference.com/w/cpp/atomic/atomic_ref). There's a bit more background on the motivation of this (from a co-author) in [this Reddit thread](https://www.reddit.com/r/cpp/comments/i8ckxr/stdatomic_ref_is_awesome/). I found about it from a mention in the talk [High-Radix Concurrent C++](https://youtu.be/75LcDvlEIYw) by Olivier Giroux (which btw is an excellent talk discussing very advanced CUDA atomics well beyond the capabilities of current compute shaders).

Thus, we have two normative references saying we can do this, SPIR-V and C++20, and a de-facto standard in spirv-cross, currently shipping applications running on Metal at scale. To my mind, the existence of atomic_ref bounds the divergence from C++. In my opinion, it would be quite reasonable for Metal to say, "in some future version of Metal, the correct way to write this is in terms of atomic_ref, but in the meantime the atomic function on a casted pointer will work, as it will be interpreted as function call that can change values in memory regardless of alias analysis."

If the answer still comes back as a no, then I will propose a revised version of my "nonatomic semantics for atomic functions" proposal, but I'm hoping we can do this in a way that's closer to existing shader translation practice.

**Comment by kvark** [comment]
Here is an update from the last office hours call.

First of all, we are talking about 3 different problems here:
  1. **Translating** from other languages (HLSL and GLSL/SPIR-V) is complicated because they don't have atomic types. No solutions here, but it can be considered to be blocked on (3) in some way.
  2. **Performance loss** when the user needs to mix atomic and non-atomic operations in the same shader. Proposed solution here (by @raphlinus, @ben-clayton, or @dneto0 ?) is to have non-atomic load and store methods exposed in WGSL for atomic types. These would map to regular load/stores in HLSL and SPIR-V, but remain atomic ops on MSL.
  3. **Bounds checking** in the generated code can't be done properly (in general!) if the driver is able to rematerialize the loading from r/w storage buffers whenever the value is needed (raised by me in https://github.com/gpuweb/gpuweb/issues/2377#issuecomment-998270169).
    - This doesn't apply to HLSL, since we don't codegen bounds checks there.
    - This weakly applies to Vulkan since we do generate some checks, but currently only R/W buffers are exposed, and Vulkan covers this with RBA. This will be a problem for R/W texture storage if/when we want to expose it - #1772. This may already be a problem with Vulkan drivers that have to generate bounds checking internally (as opposed to relying on HW features), since they are only guaranteed to do so correctly in the absence of data races... and we have data racing UB, clearly, and we decided to live with it.
    - This *fully* applies to MSL today, since we generate bounds checks. Applies to the point of being a blocker for shipping (for us), unless a solution is proposed that is confirmed to work by Apple. Resolution of this may directly affect the previous issues (1) and (2), so I believe we should focus on this one in full priority.

The important takeaway from the discussion is stressing on the fact that we are dealing with 2 kinds of UB here: unsynchronized r/w storage access on one hand (aka "data-racing UB"), and MSL compiler UB in the presence of reinterpret casts on the other hand. The latter also applies to `union` approach that @jrmuizel suggested in https://github.com/gpuweb/gpuweb/issues/2377#issuecomment-999019930. There currently is *no* proposal to solve the data-racing UB.

**Comment by raphlinus** [comment]
Great summary. A few things I'd like to add.

From my perspective, plan A is to remove the atomic types and have atomic operations on ordinary memory locations. I understand that this is technical UB in MSL, but believe that it falls within the same category of UB that we must accept (data races in particular). For HLSL and Vulkan targets I believe there is no particular issue, it's how they already work. There is reason to hope future MSL will add atomic_ref, in which case there is an entirely clean translation path. In that case, the only controversy is whether the reinterpret cast to an atomic type (as is done by spirv-cross) is a suitable workaround. From what I see, I believe it is, but I also think it's entirely fair for Apple to veto the idea if their insight into the Metal implementation suggests that it's too dangerous. The main benefit is that translation from other shader languages into WGSL becomes very straightforward.

Plan B (as kvark mentioned) is to add "non-atomic" semantics to atomic types. This has the downside of adding noise to WGSL source, but some upsides as well. Nothing is blocking this (other than the difficulty of getting it spec'ed), and if there are safety concerns it is always possible to upgrade to "relaxed," at some performance cost (I'm ok with this as long as it's a temporary condition that can be resolved by improvements in drivers). One upside is that I think we can be even finer grained in semantics, with "private" as a weaker option and "unordered" as one intermediate between "non-atomic" and "relaxed." This feature would likely be expert-only and probably most used as a translation target. A downside is that translation is more complicated: WGSL code generation of non-atomic loads and stores would need two different representations depending on whether any atomics are used on the target (globally).

Plan C (suggested by @alan-baker) is to defer this question to later and propose an extension or future version that would add the capability (I'm a little unclear whether this would be something like explicit atomic_ref, more semantics for atomic load/store, or something else). I would consider this unfortunate and a source of friction for the ecosystem, but I do see the point in getting a 1.0 out the door, as well as opening the possibility of some future MSL version solving the problem but not precluding WebGPU running on older Metal versions.

Lastly, we had a good (philosophical?) discussion about the tension between avoiding UB and being pragmatic. I remain concerned that data races can open [time-of-check time-of-use](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) vulnerabilities. As kvark described, a specific mechanism to worry about is whether this can evade bounds checks, but it is not my only concern. I think there is an empirical question here: how practical is it to *escalate* UB from a data race into a serious security vulnerability? DOS attacks are not as big a concern, and in any case would be quite difficult to prevent in general. My gut feeling is that we should be concerned, but it's possible that the practical barriers to exploiting races are significant enough that my worry is unfounded. This might be a bit off-topic of the present issue, but it's easy to see why it comes up, as I think it's worth considering the *relative* risk of mixed atomics compared with the "background radiation level" which is nonzero primarily because of data races.

**Comment by ben-clayton** [comment]
FWIW @jrprice has been doing some performance investigation work around polyfilling `i8` and `i16` support. We were able to confirm that the lack of a "non-atomic" load on the atomic was severely impacting load & store performance by experimenting with a [new `nonAtomicLoad` built-in in tint](https://dawn-review.googlesource.com/c/tint/+/75428). The perf jump for HLSL was substantial enough that we will probably want to pursue _something_ for v1.

**Comment by litherum** [comment]
(We're still investigating this internally; we haven't forgotten about it.)

**Comment by litherum** [comment]
@pow2clk @RafaelCintron @tex3d Do you think you could describe the HLSL memory model? It's difficult to try to understand what guarantees are needed to be preserved from the original source through WGSL into MSL when we don't know what guarantees the original source makes. There's only one [mention](https://github.com/microsoft/DirectX-Specs/search?q=%22memory+model%22) of the phrase "memory model" in the [DirectX-Specs](https://github.com/microsoft/DirectX-Specs) repository and it says "The details for the memory model are described outside this document."

Or, if that question is too open-ended and difficult to answer, he's an alternative question that's more specific. Consider this HLSL shader:

```HLSL
RWStructuredBuffer<int> aBuffer : register(u0);

[numthreads(2, 1, 1)]
void main() {
    aBuffer[0] = 4;
    InterlockedMax(aBuffer[0], 5); /* ideally this would just be an atomic store but it appears HLSL doesn't have such an operation */
}
```
(Assume that `aBuffer[0]` holds 0 at the time of the dispatch call.)

Is this program well-formed? Is it guaranteed to produce any specific results?

**Comment by litherum** [comment]
I guess there's a related question for GLSL. The [spec](https://www.khronos.org/registry/OpenGL/specs/gl/GLSLangSpec.4.60.html#atomic-memory-functions) doesn't list the phrase "memory model" anywhere.

```
#version 460

layout(std430, binding = 0) buffer aBuffer {
    int anArray[];
};

void main() {
    anArray[0] = 4;
    atomicExchange(anArray[0], 5);
}
```
(Again, assume that `anArray[0]` holds 0 at the time of the dispatch call.)

**Comment by raphlinus** [comment]
In the GLSL case there's a clear enough answer, not from the GLSL spec itself but from the Vulkan memory model, which I believe is considered a normative reference (especially in terms of clarifying the semantics of GLSL) from a WGSL perspective.

The nonatomic writes are a data race, therefore this program is undefined behavior (assuming a dispatch of >1 thread). If the dispatch is of a single thread, then all accesses are ordered by program order, and the buffer contains 5 at the end.

Regarding HLSL, there's no formal spec, but most of the people I talk to (the "lore") have a mental model in which the nonatomic accesses have weak semantics (in the PTX sense, also "unordered" in LLVM speak), so the buffer would be 5 in any case. Personally I would be a bit more comfortable if there was a `globallycoherent` qualifier on the buffer as well.

The question of whether to interpret nonatomics as weak is, in my opinion, orthogonal to whether atomics need a separate type. I'm not advocating for the former (though I think it is an interesting question), but I am advocating for allowing mixing of atomics and non-atomics.

**Comment by litherum** [comment]
Thanks, that's helpful.

> assuming a dispatch of >1 thread

I edited the code to add `layout(local_size_x = 2) in;`. Sorry about omitting it.

The fact that HLSL and GLSL disagree about this is concerning.

> allowing mixing of atomics and non-atomics

Just so I'm on the same page, can you demonstrate a correct (and nontrivial) program that does this in GLSL?

**Comment by raphlinus** [comment]
Also, it's not called "memory model" in the D3D specs, it's called "memory consistency model." That search would bring you to [Section 7.14 of the D3D11.3 Functional Spec](https://microsoft.github.io/DirectX-Specs/d3d/archive/D3D11_3_FunctionalSpec.htm#[7.14%20Shader%20Memory%20Consistency%20Model](https://microsoft.github.io/DirectX-Specs/d3d/archive/D3D11_3_FunctionalSpec.htm#ShaderMemoryConsistencyModel). Another highly relevant part of that is the documentation of the [sync instruction](https://microsoft.github.io/DirectX-Specs/d3d/archive/D3D11_3_FunctionalSpec.htm#inst_SYNC). These are for D3D11, but as far as I know there are no significant changes to the memory model in D3D12.

The language is very informal and operational, especially compared with something like the Vulkan memory model which has Alloy code, or the C+11 memory mode which has had over a decade of practical use and academic work. That said, I don't think it's meaningless. The relevant quote here is "The D3D11 Shader Memory Consistency Model is weak/relaxed, as generally understood in existing architectures and literature." One reasonable interpretation is that results are generally consistent with the [PTX memory consistency model](https://github.com/NVlabs/ptxmemorymodel), with some exceptions. The most interesting section is 7.14.2 which states that a race between a nonatomic write and an atomic operation in threadgroup shared memory produces "undefined results". The section goes on to state that atomics in UAV (device) memory are atomic with respect to both nonatomic loads and stores, which I think actually does answer your proposed HLSL litmus test above: this is guaranteed to terminate with 5 in the buffer.

I could be wrong in my interpretation, and am very interested in collecting examples where some implementation behaves differently than what I stated.

Just saw this as I was writing:

> Just so I'm on the same page, can you demonstrate a correct (and nontrivial) program that does this in GLSL?

Because you said nontrivial, I'm comfortable posting the [coarse path rasterizer](https://github.com/linebender/piet-gpu/blob/8bee553e6ec915c61772d25f090b7fdfb8dbb6b7/piet-gpu/shader/path_coarse.comp) from piet-gpu. That probably looks scary because it's very dense in 2D computational geometry (it's doing highly optimized flattening, conservative rasterization, and coarse winding number calculation), but the memory access patterns are reasonably straightforward. There's an `atomicExchange` on [line 251](https://github.com/linebender/piet-gpu/blob/8bee553e6ec915c61772d25f090b7fdfb8dbb6b7/piet-gpu/shader/path_coarse.comp#L251) which builds a linked list, and writes into the same buffer (through an abstraction) on line 279; these basically resolve to `memory[offset] = data;`. (there's also an atomic add on line 225, but it's not super relevant to this discussion)

What's going on here is that there is a data structure (specifically an array of `Tile`, which is an element of `Path`) which is *dynamically* allocated in the `uint memory[]` array by [an earlier pipeline stage](https://github.com/linebender/piet-gpu/blob/8bee553e6ec915c61772d25f090b7fdfb8dbb6b7/piet-gpu/shader/tile_alloc.comp). This data structure has one field which is always updated with atomic exchange. Other accesses to this memory pool are regular nonatomic reads and writes. All this is fairly conservative with respect to memory model; within a dispatch, any *given* memory address has either all atomic accesses or all nonatomic accesses.

A similar pattern is the bounding box calculation in the [path segment processor](https://github.com/linebender/piet-gpu/blob/8bee553e6ec915c61772d25f090b7fdfb8dbb6b7/piet-gpu/shader/pathseg.comp#L252), and this might actually be a clearer example. Basically, there's logic to determine whether the path segments contributing are contained entirely within one partition (workgroup), and if so the aggregated bounding box is [written directly using nonatomics](https://github.com/linebender/piet-gpu/blob/8bee553e6ec915c61772d25f090b7fdfb8dbb6b7/piet-gpu/shader/pathseg.comp#L271-L274). But if the path segments straddle a partition boundary, the partial aggregate is accumulated using [atomic min/max operations](https://github.com/linebender/piet-gpu/blob/8bee553e6ec915c61772d25f090b7fdfb8dbb6b7/piet-gpu/shader/pathseg.comp#L281-L285). This shows that the choice of whether to use atomics or no is dynamic and dependent on the input data. It would be possible to always use atomics, but it would potentially be a performance hit (I expect the nonatomic case to be a significant majority of writes).

**Comment by dneto0** [comment]
> Just so I'm on the same page, can you demonstrate a correct (and nontrivial) program that does this in GLSL?

I'm glad you're pushing on this.

In GLSL, you need to use [GL_KHR_memory_scope_semantics](https://github.com/KhronosGroup/GLSL/blob/master/extensions/khr/GL_KHR_memory_scope_semantics.txt) in order to have a precise memory model kick in.  That's what enables the use of the Vulkan memory model.

Without that, GLSL is in the same boat as HLSL: it does not have a satisfactory memory model that can answer the detailed questions you're asking.  Also Metal leans heavily on C++14, but that's incomplete in certain respects. (For example, just like OpenCL 2.0, Metal doesn't say anything useful about concurrent accesses to image memory, particularly when it might alias with memory mapped as "device".)    At a coarse grain, none of HLSL, MSL, or plain GLSL actually tell us enough to go on, by spec, because they don't have the necessary rigor in their descriptions/manuals/documentation.


Back to your example from [this comment](https://github.com/gpuweb/gpuweb/issues/2377#issuecomment-1027326677).
Let's copy that example, but add the use of the Vulkan memory model:
```glsl
#version 460
#extension GL_KHR_memory_scope_semantics : require   // Make sure we target the Vulkan memory model.

layout(std430, binding = 0) buffer aBuffer {
    int anArray[];
};

layout(local_size_x = 2) in;

void main() {
    anArray[0] = 4;                  // a non-atomic write.
    atomicExchange(anArray[0], 5);   // an atomic read-modify-write
}
```
This program is racy in the Vulkan memory model.  There are many executions where the atomic and non-atomic will race.

To make it non-racy, you have to ensure there is "enough" synchronization between the store in one invocation and the read-modify-write in the other.  Here "enough" stands for a bunch of verbiage covered by "[location-ordered](https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#memory-model-location-ordered)" in the memory model. 

A few patterns will work. The simplest is to put a control+memory barrier between the non-atomic store and the atomic read-modify-write.  *If we assume there is only one workgroup* then a regular barrier will work:

```glsl
void main() {
    anArray[0] = 4;   // a non-atomic write.

    // If only one workgroup exists, a regular barrier provides enough synchronization.
    // It induces a happens-before relation, and consequently a "location-order" relation
    // between the non-atomic write and the atomic read-modify-write.
    barrier(); 

    atomicExchange(anArray[0], 5);   // an atomic read-modify-write
}
```

If there can be more than one workgroup, then we need more.  That's because a regular GLSL control barrier only orders invocations within the same workgroup.
So we now need a handoff combining barriers and atomics to make sure events are ordered across workgroups.
See the [synchronizes-with section](https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#memory-model-synchronizes-with).    I think the most straightforward one is

> If A is a release barrier and B is an acquire barrier, then A synchronizes-with B ...

This is what it looks like.  Because the focus is on inter-workgroup communication, I simplified things and made each workgroup only have one invocation.

```
#version 460
#extension GL_KHR_memory_scope_semantics : require   // Make sure we target the Vulkan memory model.

layout(std430, binding = 0) buffer aBuffer {
    int anArray[];
    uint current_workgroup; // Helps us order operations across workgroups. Init to 0.
};

layout(local_size_x = 1) in; // one invocation per workgroup. Focus on inter-workgroup synchronization.

void main() {
    // Wait our turn.
    while(gl_WorkGroupID.x != atomicLoad(current_workgroup, gl_ScopeDevice, gl_StorageSemanticsBuffer, gl_SemanticsVolatile))
       {}

    // Synchronize with the previous workgroups.
    // Form the back two-fourths of a barrier-atomic-atomic-barrier synchronization chain.
    // We've already done the atomic-load, so now supply the barrier.
    memoryBarrierBuffer();

    anArray[0] = 4; // a non-atomic write.

    // We made workgroup size 1, so the non-atomic write is by the same
    // invocation as this atomic-read-modify-write.
    // In the Vulkan memory model, an invocation can't race with itself, so we don't
    // need a barrier, and the read-modify-write will see the value stored just above here, i.e. 4.
    const int before_value = atomicExchange(anArray[0], 5);   // an atomic read-modify-write
    // Guaranteed that before_value is 4.


    // Synchronize with the later workgroups.
    // Form the leading two-fourths of a barrier-atomic-atomic-barrier synchronization chain.
    memoryBarrierBuffer();
    atomicStore(current_workgroup, gl_WorkGroupID.x+1, gl_ScopeDevice, gl_StorageSemanticsBuffer, gl_SemanticsVolatile);

}
```

**Comment by dneto0** [comment]
> This data structure has one field which is always updated with atomic exchange. Other accesses to this memory pool are regular nonatomic reads and writes. All this is fairly conservative with respect to memory model; within a dispatch, any given memory address has either all atomic accesses or all nonatomic accesses.

Right.  In Vulkan memory model terms, @raphlinus is taking advantage of a memory barrier at the API level, to create a "[system-synchronizes-with](https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#memory-model-system-synchronizes-with)" relation between the non-atomic accesses in one dispatch and the atomic accesses in a subsequent dispatch (and vice versa).

The accesses are thus location-ordered by the case that starts with

>  Let DX and DY each be either the device domain or the host domain...

The memory barrier at API level between the different dispatches is the "domain operation" here.

And I'm sure all that is jargon that makes no sense until you spend a whole bunch of time grokking the Vulkan spec.

**Comment by dneto0** [comment]
Amusingly, the pattern in @raphlinus' application has correct well-specified behaviour in OpenCL 2.0 [pdf](https://www.khronos.org/registry/OpenCL/specs/opencl-2.0.pdf)

In particular, 3.3.6.4 Memory Ordering Rules: Host-side and Device-side Commands has the relevant clauses:

> 9. The start of a kernel-instance K global-synchronizes-with all operations in the work
items of K. Note that this includes the execution of any atomic operations by the
work items in a program using fine-grain SVM.

> 10. All operations of all work items of a kernel-instance K global-synchronizes-with the
event signaling the completion of K. Note that this also includes the execution of any
atomic operations by the work items in a program using fine-grain SVM

and the work of kernels enqueued on in-ordered command queues respect the queue order:

> 5. If commands C and C1 are enqueued in that sequence onto an in-order commandqueue, then the event (including the event implied between C and C1 due to the inorder queue) signaling C’s completion global-synchronizes-with C1. Note that in OpenCL 2.0, only a host command-queue can be configured as an in-order queue.

**Comment by m-schuetz** [comment]
> Just so I'm on the same page, can you demonstrate a correct (and nontrivial) program that does this in GLSL?

[Rendering point clouds with atomicMin.](https://github.com/m-schuetz/compute_rasterizer/blob/7e302df9945697b1e5f3bab7519771d0f08bdf1f/modules/compute_ballot_earlyDepth_dedup/render.cs#L95-L100) A cheap non-atomic early-depth test decides whether an expensive atomicMin operation is required.

**Comment by raphlinus** [comment]
> In GLSL, you need to use [GL_KHR_memory_scope_semantics](https://github.com/KhronosGroup/GLSL/blob/master/extensions/khr/GL_KHR_memory_scope_semantics.txt) in order to have a precise memory model kick in.

I believe this is not quite correct. That *GLSL* extension unlocks a whole slew of built-in functions and flags, but does not by itself change the SPIR-V. In order to do that, you also add `#pragma use_vulkan_memory_model`, which sets the VulkanMemoryModel capability, changes OpMemoryModel to Vulkan, and makes quite a range of other changes to generated SPIR-V. Key among those is adding MakePointerAvailable/Visible and NonPrivatePointer to nonatomic loads and stores to buffers with `coherent` GLSL qualifications. My understanding is that you have to set these capabilities in the SPIR-V, and *also* request vulkanMemoryModel in the [PhysicalDeviceVulkanMemoryModelFeatures](https://www.khronos.org/registry/vulkan/specs/1.3-extensions/man/html/VkPhysicalDeviceVulkanMemoryModelFeaturesKHR.html) passed in at device creation. Otherwise the spec guarantees [nothing](https://www.youtube.com/watch?v=HaoySOGlZ_U).

There's a whole other discussion to be had about what semantics to expect from GLSL/SPIR-V/Vulkan when the memory model is *not* fully opted into. I'm not sure how valuable it is to get into that, but happy to add my 2¢ if people are interested.

**Comment by dneto0** [comment]
> I believe this is not quite correct. That GLSL extension unlocks a whole slew of built-in functions and flags, but does not by itself change the SPIR-V. In order to do that, you also add #pragma use_vulkan_memory_model, which sets the VulkanMemoryModel capability, changes OpMemoryModel to Vulkan, and makes quite a range of other changes to generated SPIR-V. 

I took a look this morning.  The only change adding `#pragma use_vulkan_memory_model` does is change the memory scope on the barrier and the compare-exchange from "Device" to "QueueFamily", and add the associated capability `OpCapability VulkanMemoryModelDeviceScope`.   The material difference occurs when a single device has been partitioned into multiple sets of "queue families".  That's kind of beside the point in this discussion.
There is no difference with respect to non-private or the makepointeravailable/visible: in my sample program the availability an visibility operations are done in bulk by the barriers.

For completeness I'm attaching the disassembly of the two versions.

[handoff_with_pragma.spvasm.txt](https://github.com/gpuweb/gpuweb/files/8090137/handoff_with_pragma.spvasm.txt)
[handoff.spvasm.txt](https://github.com/gpuweb/gpuweb/files/8090138/handoff.spvasm.txt)

Definitely yes, you have to explicitly detect and opt into the Vulkan memory model at the Vulkan API level.

**Comment by dneto0** [comment]
Two things:
- In the first form of my sample [above](https://github.com/gpuweb/gpuweb/issues/2377#issuecomment-1042559545) I forgot to prevent races between the non-atomic stores in different invocations. I left a note instead of fixing it; that wasn't the kind of race we were focused on.
- The multi-workgroup sample program, I made some important assumptions about execution order and progress.  That aspect is beside the point here, again. Removing those assumptions makes the program more complicated, and distracts from the thrust of the argument.

**Comment by raphlinus** [comment]
I think I have an answer to the mystery of why David's program does not require explicit `#pragma use_vulkan_memory_model` but why my [prefix.comp](https://github.com/linebender/piet-gpu/blob/master/tests/shader/prefix.comp) example does. David's program specifies `gl_SemanticsVolatile` which maps to Volatile [memory sematics](https://www.khronos.org/registry/SPIR-V/specs/unified1/SPIRV.html#Memory_Semantics_-id-). That particular value requires VulkanMemoryModel capability, while the stuff used in my prefix sum implementation (explicit device scope, acquire/release) does not. Lesson: you have to be careful if you explicitly want Vulkan memory model semantics, otherwise you're basically depending on a compiler heuristic.

Whether volatile is appropriate is another question. My feeling is that it's not, and I think there is no live proposal to map anything in WGSL to volatile semantics (though it's been [brought up](https://github.com/gpuweb/gpuweb/issues/978)). If there were, I would argue against it. I think for the purpose of coordinating between workgroups (and even between host and workgroup, should the need arise), the concept of "coherent" covers what we need.

I will also confess I had a physical reaction to the execution and progress issues, but held back from commenting on them :)

**Comment by dneto0** [comment]
Yeah, this bug is getting super long. 

To be clear:

* Google's position is that WGSL V1 should not allow a shader to mix atomic and non-atomic operations on the same memory locations.  We are happy with how atomics are currently specified.
* Google is not proposing a "nonatomic load" operation/primitive/builtin.   We did explore the idea in the December/January timeframe but decided it was not viable for WebGPU,  because current platform APIs don't support it well enough.

This issue is still "Post-V1", which is appropriate.   My answers in the last few days are trying to explain how mixed atomic/non-atomic accesses are supposed to work.

**Comment by litherum** [comment]
I don't think we're going to be able to report before V1.

**Comment by magcius** [comment]
Since it came up today, the Direct3D specification says that doing loads of untyped atomics [without it being marked as `globallycoherent` is undefined](https://microsoft.github.io/DirectX-Specs/d3d/archive/D3D11_3_FunctionalSpec.htm#7.14.4%20Global%20vs%20Group/Local%20Coherency%20on%20Non-Atomic%20UAV%20Reads):

> If a [Compute Shader](https://microsoft.github.io/DirectX-Specs/d3d/archive/D3D11_3_FunctionalSpec.htm#ComputeShader)(18) thread in a given thread group needs to perform loads of data that was written by atomics or stores in another thread group, the UAV slot where the data resides must be tagged upon declaration in the shader as "globally coherent", so the implementation can ignore the local cache. Otherwise, this form of cross-thread group data sharing will produce undefined results.
